### PR TITLE
SNOW-1514185: Do assign the reopened channel unless kafka offsets are fully reset

### PR DESCRIPTION
**Pull request by sfc-gh-akowalczyk**
# Overview

SNOW-1514185

The change applies to channel reopen logic. The process used to consists of 3 steps:
 - reopen the channel
 - fetch the offset from the channel
 - reset internal buffer
 - rest offset in Kafka

If any of the steps 2-4 fails during `getOffsetSafeToCommitToKafka` the error will be ignored in [SnowflakeSinkTask::preCommit](https://github.com/snowflakedb/snowflake-kafka-connector/blob/15df7115dc26a2ad015f8486d6dd3f48a460e525/src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTask.java#L347-L349)

After the change the channel is assigned to the TopicPartitionChannel after all the steps are done.

## Pre-review checklist
- [ ] This change should be part of a Behavior Change Release. See [go/behavior-change](http://go/behavior-change).
- [ ] This change has passed Merge gate tests
- [ ] Snowpipe Changes
- [ ] Snowpipe Streaming Changes
- [ ] This change is TEST-ONLY
- [ ] This change is README/Javadocs only
- [ ] This change is protected by a config parameter <PARAMETER_NAME> eg `snowflake.ingestion.method`.
    - [ ] `Yes` - Added end to end and Unit Tests. 
    - [ ] `No` - Suggest why it is not param protected
- [ ] Is his change protected by parameter <PARAMETER_NAME> on the server side?
    - [ ] The parameter/feature is not yet active in production (partial rollout or PrPr, see [Changes for Unreleased Features and Fixes](http://go/ppp-prpr)).
    - [ ] If there is an issue, it can be safely mitigated by turning the parameter off. This is also verified by a test (See [go/ppp](http://go/ppp)).

### CODE REVIEW

**Review by sfc-gh-mbobowski — APPROVED** [review]

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java [hunk 1]
```diff
@@ -853,9 +853,18 @@ public long fetchOffsetTokenWithRetry() {
...
+    SnowflakeStreamingIngestChannel newChannel = reopenChannel(streamingApiFallbackInvoker);
+
+    LOGGER.warn(
+        "{} Fetching offsetToken after re-opening the channel:{}",
+        streamingApiFallbackInvoker,
+        this.getChannelNameFormatV1());
+    long offsetRecoveredFromSnowflake = fetchLatestOffsetFromChannel(newChannel);
+
     resetChannelMetadataAfterRecovery(streamingApiFallbackInvoker, offsetRecoveredFromSnowflake);
+    ;
```
Comment (sfc-gh-gjachimko) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java:865]: why the ';' here?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java [hunk 2]
```diff
@@ -958,11 +962,16 @@ private long getRecoveredOffsetFromSnowflake(
    *     snowflake.
    */
   private long fetchLatestCommittedOffsetFromSnowflake() {
+    SnowflakeStreamingIngestChannel channelToGetOffset = this.channel;
+    return fetchLatestOffsetFromChannel(channelToGetOffset);
+  }
+
+  private long fetchLatestOffsetFromChannel(SnowflakeStreamingIngestChannel channelToGetOffset) {
```
Comment (sfc-gh-gjachimko) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java:969]: channelToGetOffset -> rename to just the "channel" ?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/DirectTopicPartitionChannel.java [hunk 3]
```diff
@@ -640,9 +640,17 @@ public long fetchOffsetTokenWithRetry() {
    */
   private long streamingApiFallbackSupplier(
```
Comment (sfc-gh-gjachimko) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/DirectTopicPartitionChannel.java:641]: i'm wondering if there is a risk of having race condition here - i.e. two callers invoking this method would invoke resetChannelMetadataAfterRecovery - this method mutates underlying state, so - i'm wondering if the entire method content should be synchronized?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java (hunk 1 above)
Comment (sfc-gh-akowalczyk) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java]: just a leftover, removed

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java (hunk 2 above)
Comment (sfc-gh-akowalczyk) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java]: 👍

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/DirectTopicPartitionChannel.java (hunk 3 above)
Comment (sfc-gh-akowalczyk) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/DirectTopicPartitionChannel.java:641]: that's true, I put the channel assignment into the synchronized sections, so all properties are being modified there

File: src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java [hunk 4]
```diff
@@ -1245,4 +1247,68 @@ public void testOffsetTokenVerificationFunction() {
     Assert.assertFalse(StreamingUtils.offsetTokenVerificationFunction.verify("1", "3", "4", 3));
     Assert.assertFalse(StreamingUtils.offsetTokenVerificationFunction.verify("2", "1", "4", 3));
   }
+
+  @Test
+  public void assignANewChannelAfterTheSetupIsFullyDone() throws Exception {
+    // given
+
+    SnowflakeStreamingIngestChannel channel1 = Mockito.mock(SnowflakeStreamingIngestChannel.class);
```
Comment (sfc-gh-achyzy) [inline src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:1255]: Nit: I would like to use `FakeSnowflakeStreamingIngestChannel` or extend that implementation so we don't have to do so much mocking.

Comment (sfc-gh-akowalczyk) [reply src/test/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannelTest.java:1255]: Either way, I must modify `FakeSnowflakeStreamingIngestChannel` for a specific edge case. This means it won't be significantly different from mocking an object.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java [hunk 5]
```diff
@@ -915,6 +925,7 @@ private void resetChannelMetadataAfterRecovery(
       // Set the flag so that any leftover rows in the buffer should be skipped, it will be
       // re-ingested since the offset in kafka was reset
       needToSkipCurrentBatch = true;
+      this.channel = newChannel;
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java:928]: Is this correct? At line 904 above, we simply return if the channel offset is null, this means that we will never use the new channel if the offset is null?
